L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc

```diff
@@ -0,0 +1,178 @@
+#include <unordered_map>
+
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "DataFormats/L1TParticleFlow/interface/PFCandidate.h"
+
+#include "L1Trigger/Phase2L1ParticleFlow/src/newfirmware/deregionizer/deregionizer_input.h"
```

Comment:
I'm not sure if this structure follows CMSSW rule. Maybe @perrotta @qliphy @smuzaffar can suggest.

Reply:
Uhm, no, it looks weird,
If the header in L1Trigger/Phase2L1ParticleFlow/src/newfirmware/deregionizer/deregionizer_input.h is needed by CMSSW code in plugins, why not to put it in  L1Trigger/Phase2L1ParticleFlow/interface? Or maybe inside the /plugin area itself, if it is not needed anywhere else (I did not look at the code).
Even worse the inclusions of the cpp files in the lines below: why do you need to include the implementation file, besides the header?

Reply:
About the inclusion of the cpp files, it was done to solve compilation errors "undefined references to XXX".
I am not sure I understand the suggestion. Should I change the structure of the code and move the files in src/newfirmware/deregionizer somewhere else, or should I modify where the header files are included?

Reply:
> About the inclusion of the cpp files, it was done to solve compilation errors "undefined references to XXX".

Only visible interfaces should stay in the .h files. If anything needed from the outside world is found in the implementation file, then is should be moved to the interface

> I am not sure I understand the suggestion. Should I change the structure of the code and move the files in src/newfirmware/deregionizer somewhere else, or should I modify where the header files are included?

If deregionizer_input.h needs only to be accessed from the  L1Trigger/Phase2L1ParticleFlow/plugin area, then move it over there

If it needs to be accessed even from outside that plugin area, then move to L1Trigger/Phase2L1ParticleFlow/interface

By the way, which is the purpose of the L1Trigger/Phase2L1ParticleFlow/src/newfirmware/deregionizer area? (I repeat myself, I did not look at the code, yet). Is there any reason to move it away from main L1Trigger/Phase2L1ParticleFlow? If it acts as a separate package, then you should rather develop it in an independent (new?) package

Reply:
@gpetruc Can you please comment about the structure of L1Trigger/Phase2L1ParticleFlow/src/newfirmware and whether we can reorganize the code to respect CMS standards?

Reply:
Regarding the inclusion of the cpp files, it wasn't of my doing: in 11_1_X, the cpp files in the subdirectories were being properly compiled by scram so there was no need to include them. 
I don't know if something was changed in the BuildFile syntax in how scram works in the newer release.